### PR TITLE
Retry update_goal_state on GoalStateMismatchError

### PR DESCRIPTION
**Pull request by narrieta**

### CODE REVIEW

File: azurelinuxagent/common/protocol/extensions_goal_state.py [hunk 1]
```diff
@@ -25,8 +25,9 @@
 
 
 class GoalStateMismatchError(AgentError):
-    def __init__(self, msg):
-        super(GoalStateMismatchError, self).__init__(msg)
+    def __init__(self, message, attribute):
+        super(GoalStateMismatchError, self).__init__(message)
+        self.attribute = attribute
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/extensions_goal_state.py:30]: I extended the extension to include the name of the attribute that caused the mismatch

File: tests/protocol/mockwiredata.py [hunk 2]
```diff
@@ -38,8 +38,8 @@
         "test_ext": "ext/sample_ext-1.3.0.zip",
         "remote_access": None,
         "in_vm_artifacts_profile": None,
-        "vm_settings": "hostgaplugin/vm_settings.json",
-        "ETag": "1"
+        "vm_settings": None,
+        "ETag": None
```
Comment (narrieta) [inline tests/protocol/mockwiredata.py:42]: A large number of unit tests use the default DATA_FILE, which does not have a vm_settings.json that matches ext_conf.xml; all those test would hit the retry logic with the sleep before falling back to extensions config.

We will need matching files for all these data files, but for the moment I made DATA_FILE return 404 for vmSettings (falling back to extensions config without hitting the sleep in the retry)

File: azurelinuxagent/common/protocol/extensions_goal_state.py [hunk 3]
```diff
@@ -100,9 +100,9 @@ def compare(from_extensions_config, from_vm_settings):
         context = []  # used to keep track of the attribute that is being compared
 
         def compare_goal_states(first, second):
+            compare_attributes(first, second, "created_on_timestamp")
             compare_attributes(first, second, "activity_id")
             compare_attributes(first, second, "correlation_id")
-            compare_attributes(first, second, "created_on_timestamp")
```
Comment (kevinclark19a) [inline azurelinuxagent/common/protocol/extensions_goal_state.py:105]: Would it be worth leaving a comment here mentioning the importance of the ordering here?

Comment (narrieta) [reply azurelinuxagent/common/protocol/extensions_goal_state.py:105]: it's a good point, but all this code will be ripped off in the next iteration (which is already happening on a separate branch) and currently I do not have an environment where i can run unit tests to verify changes do this PR i will limit changes to functional/correctness issues

**Review by kevinclark19a — APPROVED** [review]

File: azurelinuxagent/common/protocol/extensions_goal_state.py (hunk 3 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/extensions_goal_state.py:105]: I had to fix some typos, so I went ahead and added the comment. Thanks for pointing it out.

**Review by kevinclark19a — APPROVED** [review]

File: tests/data/hostgaplugin/ext_conf-requested_version.xml [hunk 4]
```diff
@@ -0,0 +1,148 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (larohra) [inline tests/data/hostgaplugin/ext_conf-requested_version.xml:1]: NIT on file name - `ext_conf_requested_version.xml`

File: azurelinuxagent/common/protocol/wire.py [hunk 5]
```diff
@@ -780,70 +780,88 @@ def update_host_plugin_from_goal_state(self):
...
             #
             # The entire goal state needs to be retrieved from the WireServer (via the GoalState class), and the HostGAPlugin
-            # (via the self._fetch_vm_settings_goal_state method.)
+            # (via the self._fetch_vm_settings_goal_state method).
             #
-            # Start by fetching the goal state from the WireServer; if fetch_full_goal_state becomes True this will also
-            # update the ExtensionsConfig instance in GoalState.
+            # We fetch it in 3 parts:
+            #
+            # 1) The "main" goal state from the WireServer, which includes the incarnation, container ID, role config, and URLS
```
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: Neither "main" nor "first" fit what I was trying to express here: that I am not retrieving the entire WireServer goal state, but just the data structure that has the basics (incarnation, etc) plus pointers (urls) to the rest of the goal state,

Other suggestion?

File: azurelinuxagent/common/protocol/wire.py [hunk 6]
```diff
@@ -780,70 +780,88 @@ def update_host_plugin_from_goal_state(self):
...
+                    use_vm_settings = True
+                else:
+                    try:
+                        ExtensionsGoalState.compare(self._goal_state.extensions_config, vm_settings_goal_state)
+                        use_vm_settings = True
+                    except GoalStateMismatchError as mismatch:
+                        if not is_retry and mismatch.attribute in ("created_on_timestamp", "activity_id"):
+                            # this may be OK; a new goal state may have arrived in-between the calls to the HostGAPlugin and the WireServer;
+                            # retry one time after and then report the error if it happens again
+                            time.sleep(conf.get_goal_state_period())
```
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py:851]: actually i was thinking of harcoding it 2 seconds, but at the end went for the goal state period. I do not think that initial goal state is relevant here, since this is a retry

File: tests/data/hostgaplugin/ext_conf-requested_version.xml (hunk 4 above)
Comment (narrieta) [reply tests/data/hostgaplugin/ext_conf-requested_version.xml:1]: I am using "filename_with_underscores"-"description_with_underscores" throughout all those data files

File: azurelinuxagent/common/protocol/wire.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/common/protocol/wire.py]: You could maybe reword it as - "The essentials of a goal state (which includes the incarnation, container ID, role config, and URLS) are fetched first from Wireserver as they're needed to fetch the other stuff (extension config, etc)"?

**Review by larohra — APPROVED** [review]
LGTM

File: azurelinuxagent/common/protocol/wire.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: ok, will add that to my other iteration
